src/Event.h

```diff
@@ -0,0 +1,131 @@
+#pragma once
+
+#include <algorithm>
+#include <atomic>
+#include <cstddef>
+#include <functional>
+#include <iostream>
+#include <list>
+#include <mutex>
+#include <vector>
+
+template<typename... Args>
+class EventHandler {
+public:
+    typedef std::function<void(Args...)> HandlerFunc;
+    typedef size_t HandlerId;
+
+    explicit EventHandler(const HandlerFunc& func)
+        : m_HandlerFunc(func)
+    {
+        m_Id = ++s_IdCounter;
+    }
+
+    EventHandler(const EventHandler& handler) :
+        m_Id(handler.m_Id), m_HandlerFunc(handler.m_HandlerFunc)
+    {}
+
+    EventHandler(EventHandler&& handler)
+        : m_Id(handler.m_Id), m_HandlerFunc(handler.m_HandlerFunc)
+    {}
+
+    EventHandler& operator=(EventHandler& handler)
+    {
+        m_HandlerFunc = handler.m_HandlerFunc;
+        m_Id = handler.m_Id;
+        return *this;
+    }
+
+    EventHandler& operator=(EventHandler&& handler)
+    {
+        m_HandlerFunc = handler.m_HandlerFunc;
+        m_Id = handler.m_Id;
+        return *this;
+    }
+
+    void operator()(Args... params)
+    {
+        if (m_HandlerFunc) {
+            m_HandlerFunc(params...);
+        }
+    }
+
+    bool operator==(const EventHandler<Args...> handler)
+    {
+        return m_Id == handler.id();
+    }
+
+    HandlerId id() const { return m_Id; }
+
+private:
+    HandlerId m_Id;
```

Comment:
It's just a unint32_t, no need to typedef it to another type. Also it's not representing a type, probably better with a `uint32_t`
```suggestion
    uint32_t m_nHandlerId;
```